src/msg_recorder/recorder.cc

```diff
@@ -13,7 +13,19 @@ namespace cris::core {
 MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner)
     : Base(std::move(runner))
     , record_dir_(record_dir_prefix / RecordDirNameGenerator()) {
+    auto temp_record_dir = record_dir_prefix;
+    temp_record_dir += std::string("_snapshots");
+    record_snapshot_dir_ = temp_record_dir;
+
     std::filesystem::create_directories(record_dir_);
+    std::filesystem::create_directories(record_snapshot_dir_);
+
+    Timer t = Timer();
+    for (const auto& [type, pair] : snapshot_intervals_) {
+        GenerateSnapshot(pair.second);
+        const int& alias = pair.second;
+        t.setInterval([this, alias]() { GenerateSnapshot(alias); }, pair.first);
```

Comment:
Here it is: you captured `this`, but your threads are detached. So after destruction of the recorder, your timer thread may still access the destructed `this` (BOOM!)

There are tons of examples of how to do deamon threads correctly in our code base, bascially you create a shutdown flag and iterating on the flag.

Reply:
Or, if you REALLY wants to detach the thread while accessing `this`, try `enable_share_from_this`, and capturing the `weak_ptr` of the current object, so that the smart pointer handles the object lifecycle for you. Note that,

- You need to make sure the instance of the current class can ONLY be managed by shared_ptr. There are also lots of examples in our codebase
- Only capturing `weak_ptr` instead of `shared_ptr`, because you don't want your detached deamon owns your object (and prevent it from destructing)

Reply:
I tried with the `shutdown` flag iterating method today and the issues found are:
-  We do have more than one interval for the `sleep` function to register, so maybe generating multiple threads is a better choice? Otherwise, the current thread will be just frozen.
- I tried with the `enable_share_from_this` class property, but it didn't work out well since `MessageRecorder` is a child class.
- Also, there're two layers of `weak_ptr` check since there are two layers of lambda passing. I just am not sure it is the right way to go. I will probably call you tomorrow...

Reply:
> I tried with the `shutdown` flag iterating method today and the issues found are:
> -  We do have more than one interval for the `sleep` function to register, so maybe generating multiple threads is a better choice? Otherwise, the current thread will be just frozen.

不需要多线程，一个snapshot时间点的优先队列就够了，每次睡到下一次snapshot醒来，然后判断现在是哪种snapshot

> - I tried with the `enable_share_from_this` class property, but it didn't work out well since `MessageRecorder` is a child class.

什么child？为什么这会有关系

> - Also, there're two layers of `weak_ptr` check since there are two layers of lambda passing. I just am not sure it is the right way to go. I will probably call you tomorrow...

probably not. You don't really "check" the wp, but try converting wp to sp, and once you did it, you may keep the sp until you does not need it.